engine/runtime-compiler/src/main/java/org/enso/compiler/dump/IRDumper.java

```diff
@@ -681,70 +692,188 @@ private void addNode(GraphVizNode node) {
+  private void createEdge(
+      Object from,
+      Object to,
+      String label,
+      Function<JSVizEdge.Builder, JSVizEdge.Builder> bldrFunc) {
     assert !(from instanceof String);
     assert !(to instanceof String);
-    assert !(from instanceof GraphVizNode);
-    assert !(to instanceof GraphVizNode);
+    assert !(from instanceof JSVizNode);
+    assert !(to instanceof JSVizNode);
     var fromId = Utils.id(from);
     var toId = Utils.id(to);
-    var edge = GraphVizEdge.newEdgeWithAttributes(fromId, toId, label, attrs);
-    var nodesContainsFrom = nodes.stream().anyMatch(node -> node.id().equals(fromId));
-    var nodesContainsTo = nodes.stream().anyMatch(node -> node.id().equals(toId));
-    assert nodesContainsFrom
+    var fromNode = nodes.stream().filter(node -> node.id().equals(fromId)).findFirst();
+    var toNode = nodes.stream().filter(node -> node.id().equals(toId)).findFirst();
+    assert fromNode.isPresent()
         : "Node " + fromId + " not found. You must first create it before creating an edge from it";
-    assert nodesContainsTo
+    assert toNode.isPresent()
         : "Node " + toId + " not found. You must first create it before creating an edge to it";
+
+    if (!(fromNode.get().level() < toNode.get().level())) {
+      throw new AssertionError(
+          "The 'from' node must be at a lower level than the 'to' node: "
+              + "fromNode: "
+              + fromNode.get()
+              + ", toNode: "
+              + toNode.get());
+    }
+    var edgeBldr = new JSVizEdge.Builder();
+    edgeBldr.fromId(fromId).toId(toId).label(label);
+    edgeBldr = bldrFunc.apply(edgeBldr);
+    var edge = edgeBldr.build();
     var edgeAlreadyExists = edges.stream().anyMatch(e -> e.equals(edge));
     if (!edgeAlreadyExists) {
       edges.add(edge);
     }
   }
 
   private void createEdge(Object from, Object to, String label) {
-    createEdge(from, to, label, Map.of());
+    createEdge(from, to, label, Function.identity());
   }
 
-  /** Dump all the nodes and edges definitions into the GraphViz format. */
-  private void dumpGraph() {
-    write("digraph {");
-    write(System.lineSeparator());
-    for (GraphVizNode node : nodes) {
-      var nodeRepr = node.toGraphViz();
-      write(nodeRepr);
-      write(System.lineSeparator());
-    }
-    for (GraphVizEdge edge : edges) {
-      var containsFromNode = nodes.stream().anyMatch(node -> node.id().equals(edge.from()));
-      var containsToNode = nodes.stream().anyMatch(node -> node.id().equals(edge.to()));
-      assert containsFromNode;
-      assert containsToNode;
-      var edgeRepr = edge.toGraphViz();
-      write(edgeRepr);
-      write(System.lineSeparator());
+  /**
+   * Dump all the nodes and edges definitions into the JSviz format, along with the rest of the html
+   * - creates a valid HTML file.
+   */
+  private void dumpGraph(String moduleName) {
```

Comment:
This is the gist of the change. Just output a single HTML instead of GraphViz file.